llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -7283,31 +7283,35 @@ static bool isGuaranteedNotToBeUndefOrPoison(
   // BB1:
   //   CtxI ; V cannot be undef or poison here
   auto *Dominator = DNode->getIDom();
-  while (Dominator) {
-    auto *TI = Dominator->getBlock()->getTerminator();
-
-    Value *Cond = nullptr;
-    if (auto BI = dyn_cast_or_null<BranchInst>(TI)) {
-      if (BI->isConditional())
-        Cond = BI->getCondition();
-    } else if (auto SI = dyn_cast_or_null<SwitchInst>(TI)) {
-      Cond = SI->getCondition();
-    }
+  // This check is purely for compile time reasons: we can skip the IDom walk
+  // if what we are checking for includes undef and the value is not a boolean.
+  if (!includesUndef(Kind) || V->getType()->isIntOrIntVectorTy(1))
+    while (Dominator) {
+      auto *TI = Dominator->getBlock()->getTerminator();
+
+      Value *Cond = nullptr;
+      if (auto BI = dyn_cast_or_null<BranchInst>(TI)) {
+        if (BI->isConditional())
+          Cond = BI->getCondition();
+      } else if (auto SI = dyn_cast_or_null<SwitchInst>(TI)) {
```

Comment:
Hm, I just realized that the switch case is applicable to all integer types, unlike the branch case, which only works for i1.

Maybe rather than trying to skip this code entirely we should add a limit on how many idoms we inspect? For practical purposes, limiting the walk to 10 idoms is probably quite sufficient.

Reply:
Changing the skip check from `V->getType()->isIntOrIntVectorTy(1)` to `V->getType()->isIntOrIntVectorTy()` should hopefully cover all for branches and selects.

Reply:
> err, yes, thanks :) Actually, I still this skipping the code entirely (apart from adding the limit to idoms) might be useful for non-integer values (for example: trying to prove noundef for pointer loads across phis, which we do downstream). I checked how many lit tests in transform directory can skip this code entirely (added an assert if the Value is non-integer when we reach this IDom walk) : ~100 tests and some of them in LoopVectorizer are pointer Values.

The more interesting question is probably what percentage of queries use non-integer values, and I'd suspect it would be very low ... but I guess it doesn't hurt either.

You should be able to use isIntegerTy() rather than isIntOrIntVectorTy() as vectors cannot be used as branch/switch conditions.